src/hotspot/cpu/x86/universalUpcallHandler_x86_64.cpp

```diff
@@ -141,3 +149,677 @@ address ProgrammableUpcallHandler::generate_upcall_stub(jobject rec, jobject jab
+  for (int i = 0; i < conv._args_length; i++) {
+    VMReg reg = conv._arg_regs[i];
+    if (reg->is_Register()) {
+      __ movptr(reg->as_Register(), Address(rsp, load_offset));
+      load_offset += 8;
+    } else if (reg->is_XMMRegister()) {
+      // Java API doesn't support vector args
+      __ movdqu(reg->as_XMMRegister(), Address(rsp, load_offset));
+      load_offset += 16;
+    }
+    // do nothing for stack
+  }
+  __ block_comment("} restore_native_args ");
+}
+
+static bool is_valid_XMM(XMMRegister reg) {
+  return reg->is_valid() && (UseAVX >= 3 || (reg->encoding() < 16)); // why is this not covered by is_valid()?
+}
+
+// for callee saved regs, according to the caller's ABI
+static int compute_reg_save_area_size(const ABIDescriptor& abi) {
+  int size = 0;
+  for (Register reg = as_Register(0); reg->is_valid(); reg = reg->successor()) {
+    if (reg == rbp || reg == rsp) continue; // saved/restored by prologue/epilogue
+    if (!abi.is_volatile_reg(reg)) {
+      size += 8; // bytes
+    }
+  }
+
+  for (XMMRegister reg = as_XMMRegister(0); is_valid_XMM(reg); reg = reg->successor()) {
+    if (!abi.is_volatile_reg(reg)) {
+      if (UseAVX >= 3) {
+        size += 64; // bytes
+      } else if (UseAVX >= 1) {
+        size += 32;
+      } else {
+        size += 16;
+      }
+    }
+  }
+
+  return size;
+}
+
+static int compute_arg_save_area_size(const CallRegs& conv) {
+  int result_size = 0;
+  for (int i = 0; i < conv._args_length; i++) {
+    VMReg reg = conv._arg_regs[i];
+    if (reg->is_Register()) {
+      result_size += 8;
+    } else if (reg->is_XMMRegister()) {
+      // Java API doesn't support vector args
+      result_size += 16;
+    }
+    // do nothing for stack
+  }
+  return result_size;
+}
+
+static void preserve_callee_saved_registers(MacroAssembler* _masm, const ABIDescriptor& abi, int reg_save_area_offset) {
+  // 1. iterate all registers in the architecture
+  //     - check if they are volatile or not for the given abi
+  //     - if NOT, we need to save it here
+  // 2. save mxcsr (?)
+
+  int offset = reg_save_area_offset;
+
+  __ block_comment("{ preserve_callee_saved_regs ");
+  for (Register reg = as_Register(0); reg->is_valid(); reg = reg->successor()) {
+    if (reg == rbp || reg == rsp) continue; // saved/restored by prologue/epilogue
+    if (!abi.is_volatile_reg(reg)) {
+      __ movptr(Address(rsp, offset), reg);
+      offset += 8;
+    }
+  }
+
+  for (XMMRegister reg = as_XMMRegister(0); is_valid_XMM(reg); reg = reg->successor()) {
+    if (!abi.is_volatile_reg(reg)) {
+      if (UseAVX >= 3) {
+        __ evmovdqul(Address(rsp, offset), reg, Assembler::AVX_512bit);
+        offset += 64;
+      } else if (UseAVX >= 1) {
+        __ vmovdqu(Address(rsp, offset), reg);
+        offset += 32;
+      } else {
+        __ movdqu(Address(rsp, offset), reg);
+        offset += 16;
+      }
+    }
+  }
+  __ block_comment("} preserve_callee_saved_regs ");
+
+  // TODO mxcsr
```

Comment:
Anything left to do with mxcsr?

Reply:
I guess this slipped through with the initial PR.

JNI code loads the default value of mxcsr here and saves/restores the incoming native value: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/stubGenerator_x86_64.cpp#L265-L298

However, it does nothing on Windows, while the Windows x64 ABI also defines mxcsr to have a non-volatile portion (i.e. that should be saved and restored): https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-160#mxcsr

I think this made me a bit unsure about the need for saving and restoring mxcsr, and wanted to find an actual test case that causes a problem, rather than just copy-pasting the code. But, at this time I haven't gotten to that yet.

Reply:
Thanks for the pointers. It would be safer for now to align the behavior with the call stub.